docs/proposal/circuit_breaker.md

```diff
@@ -0,0 +1,98 @@
+---
+title: Circuit Breaker
+authors:
+- "@Okabe-Rintarou-0" # Authors' GitHub accounts here.
+reviewers:
+- "@supercharge-xsy"
+- "@hzxuzhonghu"
+- "@nlwcy"
+- TBD
+approvers:
+- "@robot"
+- TBD
+
+creation-date: 2024-05-29
+
+---
+
+## Add circuit breaker function in Kmesh
+
+### Summary
+
+Roadmap:
+
++ Support circuit breaker.
++ Support outlier detection.
++ Add sufficient unit tests.
+
+### Motivation
+
+The circuit breaker mechanism is typically used to prevent the spread of failures between services, safeguarding system stability and avoiding system crashes or cascading failures caused by a large number of requests. Kmesh currently does not implement the circuit breaker mechanism.
+
+Common scenarios that trigger circuit breakers include:
++ High error rate in the service
++ High latency in the service
++ Exhaustion of service resources
++ Service unavailability
+
+#### Goals
+
++ Support circuit breaker. Kmesh should be able to parse the circuit breaker config from xds stream and enable corresponding circuit breakers.
++ Support circuit breaker. Kmesh should be able to parse the outlier detection config from xds stream and support outlier detection.
++ Add sufficient unit tests to test the correctness of the functions.
+
+### Design Details
+
+#### Implement the state machine of a circuit breaker
+
+![](./pics/circuit_breaker_state_machine.png)
+
+We can implement a circuit breaker mechanism to ensure system stability and reliability while minimizing access to faulty services. First, we need to implement a circuit breaker with three states: closed, open, and half-open.
+
+In the closed state, all requests pass through normally and are sent to the target service, even if occasional failures occur, without triggering the circuit breaker. However, when the number of failed requests exceeds a preset threshold, the circuit breaker quickly switches to the open state. In this state, all requests immediately fail, avoiding sending requests to a service known to be failing, thus reducing the system's burden.
+
+In the open state, the circuit breaker starts a timeout timer. After the timeout period, the circuit breaker transitions to the half-open state. In the half-open state, the system allows a small number of requests to go through to test whether the target service has recovered. If these requests succeed, the circuit breaker closes again, allowing the service to handle more requests. If the requests fail, the circuit breaker reopens, maintaining protection for the system against the potentially faulty service.
```

Comment:
It would be better to add some implementation details, such as:
1. How Is the Timer Logic Implemented
2. Specifies the BPF hook point that triggers circuit breaker

Reply:
Sure, it's a rough outline of the current proposal. I need more follow-up discussions and code review to obtain a more detailed design. I will keep working on it.

Reply:
timer logic might be implemented by ebpf timer: https://ebpf-docs.dylanreimerink.nl/linux/concepts/timers/

here's an example:

```c
// Initialize the timer.
// First 4 bits of 'flags' specify clockid.
// Only CLOCK_MONOTONIC, CLOCK_REALTIME, CLOCK_BOOTTIME are allowed.
long bpf_timer_init(struct bpf_timer *timer, struct bpf_map *map, int flags);

// Configure the timer to call 'callback_fn' static function.
long bpf_timer_set_callback(struct bpf_timer *timer, void *callback_fn);

// Arm the timer to expire 'nsec' nanoseconds from the current time.
long bpf_timer_start(struct bpf_timer *timer, u64 nsec, u64 flags);

// Cancel the timer and wait for callback_fn to finish if it was running.
long bpf_timer_cancel(struct bpf_timer *timer);

Here is how BPF program might look like:
struct map_elem {
    int counter;
    struct bpf_timer timer;
};

struct {
    __uint(type, BPF_MAP_TYPE_HASH);
    __uint(max_entries, 1000);
    __type(key, int);
    __type(value, struct map_elem);
} hmap SEC(".maps");

static int timer_cb(void *map, int *key, struct map_elem *val);
/* val points to particular map element that contains bpf_timer. */

SEC("fentry/bpf_fentry_test1")
int BPF_PROG(test1, int a)
{
    struct map_elem *val;
    int key = 0;

    val = bpf_map_lookup_elem(&hmap, &key);
    if (val) {
        bpf_timer_init(&val->timer, &hmap, CLOCK_REALTIME);
        bpf_timer_set_callback(&val->timer, timer_cb);
        bpf_timer_start(&val->timer, 1000 /* call timer_cb2 in 1 usec */, 0);
    }
}
```

Reply:
Great~ As described, bpf_timer does not have ctx context information, so the definition of the input parameter map of timer_callback is very important. 
The key of map needs to be able to identify different circuit breakers, and the context information of the circuit breaker event needs to be recorded in the vaule of map.